File: src/verifier/summary.py
"""ADR-0022 Decisions 1 and 3 — the contribution summary, recomputed.

Kept apart from `decision.py` for the reason `electorate.py` and
`equivocation.py` are: these are the parity twins of TypeScript
`src/instance/src/hub/summary.ts`, mirrored from the ADR's own algorithm rather
than from the TypeScript internals, and an auditor comparing the two
implementations should find the pair self-contained in one file.

The defect this closes is campaign 10's through-line: **a sum is only as
recomputable as its input set is agreed, and the protocol did not name sets.**
Every mechanism before P7 answers a question about an event you can point at.
`afp:ContributionSummary` was the first object whose subject is a *boundary*,
and it shipped with a period selected on self-asserted clocks, an input scope
that four honest members resolve four different ways, and an `afp:inputHash`
whose preimage no document ever defined — a digest that reads exactly like a
check and is not one.

**Compatibility (binding).** Every function here fires only on an
`afp:ContributionSummary`, and no bundle written before this ADR carries one,
so nothing below can fail a pre-ADR export.
"""

from __future__ import annotations

from decision import afp_object, enrolled_instances, instant_millis, settlement_payload
from proof import digest_of

#: Closed registries (ADR-0022 W0.4). An unrecognised form fails; it never falls
#: through to a default, because a frame nobody can resolve is precisely the
#: unfalsifiable claim Decision 1 exists to remove.
PERIOD_FORMS = ("hub-observed",)
SPLIT_FORMS = ("declared-shares",)
VISIBILITY_CLASSES = ("public", "hub", "parties", "internal")


def summary_object(activity: dict) -> dict | None:
    """The `afp:ContributionSummary` payload of a `Create`, or `None`."""
    return afp_object(activity, "afp:ContributionSummary")


def frame_of(summary: dict) -> dict | None:
    """`afp:frame`, or `None` when the summary declares none — which is itself
    the finding, not a shape to work around."""
    frame = summary.get("afp:frame")
    return frame if isinstance(frame, dict) else None


def frame_is_known(frame: dict) -> list[str]:
    """Every reason this frame cannot be resolved, or `[]`.

    Returned as a list rather than a bool because a frame with three problems
    should say three things: a reader fixing one at a time otherwise gets a new
    failure each run, which is the worst possible way to learn a schema.
    """
    problems: list[str] = []

    period = frame.get("afp:periodRule")
    if not isinstance(period, dict):
        problems.append("afp:periodRule is absent")
    elif period.get("afp:form") not in PERIOD_FORMS:
        problems.append(f"unknown afp:periodRule form {period.get('afp:form')!r}")
    else:
        for field in ("afp:hub", "afp:from", "afp:to"):
            if not isinstance(period.get(field), str):
                problems.append(f"afp:periodRule is missing {field}")

    scope = frame.get("afp:inputScope")
    if not isinstance(scope, dict):
        problems.append("afp:inputScope is absent — a summary that does not say what it could read is not recomputable")
    else:
        classes = scope.get("afp:visibility")
        if not isinstance(classes, list) or not classes:
            problems.append("afp:inputScope declares no afp:visibility classes")
        else:
            unknown = [c for c in classes if c not in VISIBILITY_CLASSES]
            if unknown:
                problems.append("unknown visibility class(es): " + ", ".join(map(str, unknown)))

    split = frame.get("afp:splitRule")
    if not isinstance(split, dict) or split.get("afp:form") not in SPLIT_FORMS:
        problems.append(f"unknown afp:splitRule form {(split or {}).get('afp:form')!r}")

    if not isinstance(frame.get("afp:vocabulary"), str):
        problems.append("afp:vocabulary is absent — a roll-up over historical inputs must say which vocabulary read them")

    return problems


def hub_chain_slice(pool: list[dict], hub: str, start: str, end: str) -> list[dict] | None:
    """The hub's own activities in `(start, end]`, in chain order — or `None`
    when the interval does not resolve in this pool.

    Walked backwards from `end` along `afp:prevActivity`, never sorted by
    `published`: the chain is the hub's own assertion of order, and relying on
    it rather than on wall-clock is the whole of Decision 1's period rule.

    `None` and `[]` are different answers and the caller must keep them apart —
    an unresolvable interval that read as an empty period would sum nothing and
    pass, which is the vacuous shape this repository fails checks for.
    """
    by_digest = {digest_of(a): a for a in pool if a.get("actor") == hub}
    cursor = by_digest.get(end)
    if cursor is None:
        return None
    walked: list[dict] = []
    while cursor is not None:
        if digest_of(cursor) == start:
            return list(reversed(walked))
        walked.append(cursor)
        prev = cursor.get("afp:prevActivity")
        cursor = by_digest.get(prev) if isinstance(prev, str) else None
    return list(reversed(walked)) if start == "" else None


def credit_of(result: dict) -> list[tuple[str, int, int]]:
    """`[(agent, share, denominator)]` for one Result — ADR-0022 Decision 2's
    arithmetic applied. A single-author Result is the whole of itself."""
    attributed = result.get("attributedTo")
    split = result.get("afp:contributionSplit")
    if isinstance(attributed, list) and len(attributed) > 1 and isinstance(split, dict):
        denominator = sum(v for v in split.values() if isinstance(v, int) and not isinstance(v, bool))
        if denominator <= 0:
            return []
        return [(a, s, denominator) for a, s in sorted(split.items()) if isinstance(s, int)]
    if isinstance(attributed, str):
        return [(attributed, 1, 1)]
    if isinstance(attributed, list) and attributed and isinstance(attributed[0], str):
        return [(attributed[0], 1, 1)]
    return []


def _lcm(values: list[int]) -> int:
    from math import gcd

    lcm = 1
    for value in sorted(values):
        if value <= 0:
            return 0
        lcm = lcm // gcd(lcm, value) * value
    return lcm


def recompute(pool: list[dict], frame: dict, hub_actor: str) -> dict | None:
    """Decision 1 and 3's arithmetic: `{denominator, credited, unreadable,
    inputs}` for the period the frame declares, or `None` when the interval does
    not resolve here.

    The join is protocol-level and reads no naming convention: the chain slice
    gives the settlements the hub sequenced, each settlement names its task, the
    task's own `Announce` gives the thread, and the Results on that thread are
    the work. A deployment that numbers its correlation ids differently
    recomputes identically.
    """
    # Deduplicated by digest before anything is counted. Not defensive tidying:
    # a joint replay's merged pool legitimately holds the same Result twice —
    # the author's own copy and the verbatim copy a counterparty received across
    # the boundary (ADR-0009) — and summing the pool as given credits that work
    # twice, a wrong number arrived at from an entirely correct record.
    pool = list({digest_of(a): a for a in pool}.values())

    period = frame["afp:periodRule"]
    slice_ = hub_chain_slice(pool, period["afp:hub"], period["afp:from"], period["afp:to"])
    if slice_ is None:
        return None

    scope = set(frame["afp:inputScope"]["afp:visibility"])
    announces: dict[str, dict] = {}
    for activity in pool:
        task = afp_object(activity, "afp:Task")
        if task is not None and isinstance(task.get("id"), str):
            announces[task["id"]] = activity

    instances = enrolled_instances(hub_actor, pool, instant_millis(None) or 2**63)
    def operator_of(agent: str) -> str:
        return instances.get(agent, agent)

    rows: list[tuple[str, int, int]] = []
    unreadable: dict[str, int] = {}
    inputs: set[str] = set()

    for activity in slice_:
        # `settlement_payload`, not `afp_object`: an `afp:Settlement` travels as a
        # BARE afp-typed activity carrying its payload in `object`, so
        # `afp_object` matches the envelope on its first branch and hands back
        # an activity with no `afp:task` on it. The two implementations
        # disagreed here on the first run, which is the entire reason they are
        # written from the ADR rather than from each other.
        settlement = settlement_payload(activity)
        if settlement is None:
            continue
        inputs.add(digest_of(activity))
        announce = announces.get(str(settlement.get("afp:task", "")))
        thread = announce.get("context") if isinstance(announce, dict) else None
        results = [
            a
            for a in pool
            if thread is not None
            and a.get("context") == thread
            and afp_object(a, "afp:Result") is not None
            and a.get("afp:visibility") in scope
        ]
        if not results:
            # The hub sequenced a settlement, so the work happened; this
            # computer cannot read the Result that says who did it. Counted,
            # never estimated, never silently skipped (W0.2) — attributed
            # through the settlement's own entries, which the hub published and
            # every member can read.
            settles = settlement.get("afp:settles")
            actors = (
                [str(e.get("actor")) for e in settles if isinstance(e, dict) and isinstance(e.get("actor"), str)]
                if isinstance(settles, list)
                else []
            )
            for operator in sorted({operator_of(a) for a in (actors or ["unattributed"])}):
                unreadable[operator] = unreadable.get(operator, 0) + 1
            continue
        for result in results:
            inputs.add(digest_of(result))
            rows.extend(credit_of(afp_object(result, "afp:Result") or {}))

    denominator = _lcm([row[2] for row in rows]) if rows else 1
    credited: dict[str, int] = {}
    for agent, share, own in rows:
        operator = operator_of(agent)
        credited[operator] = credited.get(operator, 0) + share * denominator // own

    return {
        "denominator": denominator,
        "credited": credited,
        "unreadable": unreadable,
        "inputs": sorted(inputs),
    }


def check_summary_frame(report, activity: dict) -> None:
    """ADR-0022 Decision 1 / W3 V2 — the frame is present and every form in it
    is one this verifier can resolve. Per domain: a frame is a shape claim about
    the summary's own bytes and needs nothing from anybody else."""
    summary = summary_object(activity)
    if summary is None:
        return
    label = summary.get("id", activity.get("id", "<no id>"))
    frame = frame_of(summary)
    if frame is None:
        report.record(
            f"contribution: {label} summary frame is a known form",
            False,
            "no afp:frame — a summary that does not declare which events it summed, over which "
            "window, under which visibility classes, is a number in a signed envelope (ADR-0022 "
            "Decision 1)",
        )
        return
    problems = frame_is_known(frame)
    report.record(
        f"contribution: {label} summary frame is a known form",
        not problems,
        "" if not problems else "; ".join(problems) + " (ADR-0022 Decision 1)",
    )


def check_summary_arithmetic(report, bundles: list[dict | None]) -> None:
    """ADR-0022 Decisions 1 and 3 / W3 V3 and V4 — **replay-wide and
    three-valued**, for the reason ADR-0021's V4 already established: the
    evidence belongs to other parties.

    A summary sums four operators' Results and a hub's chain. The computer's own
    bundle holds its half; the rest lives in the bundles beside it. Folding this
    per domain would fail every non-host bundle in any real case file, so the
    pool is merged, and a replay that cannot resolve the declared interval at
    all records **unresolvable** rather than a failure — which is not the same
    as recording a pass, and prints in the census as its own count.

    V4 is the half campaign 10 was actually about. A summary computed over a
    partial view is legitimate — 07's classes are doing their job, and ADR-0013
    is right to 404 a competitor — but a summary that sums a partial view and
    says nothing about it is indistinguishable from one that summed everything,
    which is how two honest members disagreed with no way to tell an entitlement
    gap from a fraud.
    """
    present = [b for b in bundles if b is not None]
    pool: list[dict] = []
    for bundle in present:
        pool.extend(bundle["activities"])

    for bundle in present:
        domain = bundle["path"].name or str(bundle["path"])
        prefix = f"[{domain}] " if len(present) > 1 else ""
        for activity in bundle["activities"]:
            summary = summary_object(activity)
            if summary is None:
                continue
            label = summary.get("id", activity.get("id", "<no id>"))
            frame = frame_of(summary)
            if frame is None or frame_is_known(frame):
                continue  # V2 already failed this one by name; do not fail it twice

            hub_actor = str(summary.get("afp:hub") or frame["afp:periodRule"]["afp:hub"])
            recomputed = recompute(pool, frame, hub_actor)

            v3 = f"{prefix}contribution: {label} input hash matches the set its frame declares"
            v4 = f"{prefix}contribution: {label} unreadable inputs are counted, not dropped"
            if recomputed is None:
                detail = (
                    "unresolvable: this replay does not carry the hub chain the frame's period "
                    "names, so the summed set cannot be rebuilt from it (ADR-0022)"
                )
                report.record(v3, True, detail)
                report.record(v4, True, detail)
                continue

            declared_hash = summary.get("afp:inputHash")
            recomputed_hash = digest_of(sorted(recomputed["inputs"]))
            report.record(
                v3,
                declared_hash == recomputed_hash,
                ""
                if declared_hash == recomputed_hash
                else f"afp:inputHash {declared_hash!r} is not the digest of the sorted activity "
                f"digests the frame selects ({len(recomputed['inputs'])} of them) — the field "
                f"exists so a second party can confirm it summed the same things (ADR-0022 Decision 3)",
            )

            declared = summary.get("afp:unreadable")
            declared_counts = (
                {str(e.get("afp:operator")): e.get("afp:count") for e in declared if isinstance(e, dict)}
                if isinstance(declared, list)
                else None
            )
            if declared_counts is None:
                report.record(
                    v4,
                    False,
                    "no afp:unreadable census — a summary states what it could not read, or a "
                    "partial view is indistinguishable from a complete one (ADR-0022 Decision 1)",
                )
                continue
            missing = {
                operator: count
                for operator, count in recomputed["unreadable"].items()
                if declared_counts.get(operator) != count
            }
            report.record(
                v4,
                not missing,
                ""
                if not missing
                else "declared census disagrees with the recomputed one for "
                + ", ".join(f"{operator} (recomputed {count})" for operator, count in sorted(missing.items()))
                + " (ADR-0022 Decision 1)",
            )

File: src/instance/src/hub/summary.ts
/**
 * ADR-0022 Decisions 1 and 3 — the contribution summary, and the frame that
 * makes it recomputable.
 *
 * Kept apart from `hub.ts` for the reason `electorate.ts` and `equivocation.ts`
 * are: these are pure functions over signed activities, they are the parity
 * twins of `src/verifier/summary.py`, and an auditor comparing the two
 * implementations should find each half self-contained in one file.
 *
 * **A summary is not a hub artifact.** 04 is deliberate that `afp:computedBy`
 * is "a field, not a privileged role" — anyone holding the record may compute
 * one, which is the entire reason the object is worth having. So nothing here
 * touches the hub's database or its keys: `computeContribution` takes a pool of
 * activities and returns numbers, and any member can call it over whatever it
 * can read.
 *
 * What campaign 10 found, and what this file answers: **a sum is only as
 * recomputable as its input set is agreed.** Two honest members recomputing the
 * same quarter got different numbers and neither had done anything wrong — one
 * of them simply could not read a fifth of the work (07's visibility classes,
 * ADR-0013's read gate, both behaving correctly). So the summary declares its
 * frame — which events, over which window, seen under which classes — and
 * counts what it could not read rather than quietly summing without it.
 */

import type { JsonValue } from "../crypto/jcs.ts";
import { digestOf } from "../crypto/proof.ts";
import type { Envelope, Visibility } from "../ap/activities.ts";
import { AFP_CONTEXTS } from "../ap/documents.ts";
import { lcmOf } from "./weights.ts";

type Activity = { [key: string]: JsonValue };

/**
 * ADR-0022 Decision 1: the four things a second party needs in order to sum the
 * same set. A closed registry per field — an unrecognised form fails rather
 * than falling through to a default (W0.4), because a frame nobody can resolve
 * is exactly the unfalsifiable claim this decision exists to remove.
 */
export interface SummaryFrame {
  /**
   * `hub-observed`: the period is the half-open interval of the hub's own
   * chain between two of its activity digests — `(from, to]`.
   *
   * Wall-clock `afp:period` stays on the summary as narration and is never
   * what the arithmetic selects on. `published` is self-asserted (campaign 7's
   * finding 45), so a task that straddles a quarter boundary would otherwise
   * belong to whichever quarter its own author said it did. The hub's chain is
   * the one order every member observes identically and no member can move.
   */
  periodRule: { form: "hub-observed"; hub: string; from: string; to: string };
  /** The visibility classes the computer summed over — what it was entitled to read. */
  inputScope: { visibility: readonly Visibility[] };
  /** `declared-shares`: credit follows `afp:contributionSplit` (Decision 2). */
  splitRule: { form: "declared-shares" };
  /** The vocabulary version this was computed under (finding 74). */
  vocabulary: string;
}

export interface ContributionTotals {
  /** The common denominator every credit is expressed over — integers only. */
  denominator: number;
  /** operator → credited numerator. Divide by `denominator` for tasks-equivalent. */
  credited: Record<string, number>;
  /**
   * operator → count of settled tasks whose Result this computer could not
   * read. Known to have happened (the hub sequenced the settlement and the
   * award names the performer) and impossible to credit in detail.
   */
  unreadable: Record<string, number>;
  /** The digests summed, sorted — Decision 3's preimage. */
  inputs: string[];
}

function objectOf(activity: Activity): Record<string, JsonValue> | null {
  const object = activity.object;
  if (typeof object !== "object" || object === null || Array.isArray(object)) return null;
  return object as Record<string, JsonValue>;
}

function payloadOf(activity: Activity, type: string): Record<string, JsonValue> | null {
  const object = objectOf(activity);
  return object && object.type === type ? object : null;
}

/**
 * The hub's own activities in `(from, to]`, in chain order.
 *
 * Walked backwards from `to` along `afp:prevActivity` rather than sorted by
 * `published`: the chain is the hub's own assertion of order and it is the
 * thing this rule exists to rely on. A `from` that never appears means the
 * interval is unresolvable, which the caller must treat as such — never as an
 * empty period, which would silently sum nothing and pass.
 */
export function hubChainSlice(pool: readonly Activity[], hub: string, from: string, to: string): Activity[] | null {
  const byDigest = new Map<string, Activity>();
  for (const activity of pool) {
    if (activity.actor === hub) byDigest.set(digestOf(activity), activity);
  }
  const slice: Activity[] = [];
  let cursor = byDigest.get(to);
  if (!cursor) return null;
  while (cursor) {
    const digest = digestOf(cursor);
    if (digest === from) return slice.reverse();
    slice.push(cursor);
    const prev = cursor["afp:prevActivity"];
    cursor = typeof prev === "string" ? byDigest.get(prev) : undefined;
  }
  // Ran off the head of the chain without meeting `from`: the interval does not
  // resolve in this pool.
  return from === "" ? slice.reverse() : null;
}

/**
 * ADR-0022 Decision 2's arithmetic, applied: `agent → share` over a common
 * denominator. A single-author Result is the whole of itself.
 */
export function creditOf(result: Record<string, JsonValue>): { agent: string; share: number; denominator: number }[] {
  const attributed = result.attributedTo;
  const split = result["afp:contributionSplit"];
  if (Array.isArray(attributed) && attributed.length > 1 && split && typeof split === "object" && !Array.isArray(split)) {
    const shares = split as Record<string, number>;
    const denominator = Object.values(shares).reduce((sum, share) => sum + share, 0);
    return Object.entries(shares).map(([agent, share]) => ({ agent, share, denominator }));
  }
  const sole = typeof attributed === "string" ? attributed : Array.isArray(attributed) ? String(attributed[0]) : "";
  return sole ? [{ agent: sole, share: 1, denominator: 1 }] : [];
}

/**
 * ADR-0022 Decisions 1 and 3 — recompute a period's contribution from the pool.
 *
 * The join is protocol-level, not conventional: the hub's chain slice gives the
 * settlements it sequenced in the period; each settlement names its task; the
 * task's own `Announce` gives the thread; the Results on that thread are the
 * work. Nothing here reads a naming convention, so a deployment that numbers
 * its correlation ids differently recomputes identically.
 *
 * `resolveOperator` maps an agent to the instance that operates it — the same
 * Enroll-trail fold ADR-0005 buckets weights by. An agent it cannot resolve is
 * credited under its own URL rather than dropped: an unknown operator is a
 * visible oddity, and a dropped credit is an invisible one.
 */
export function computeContribution(
  poolIn: readonly Activity[],
  frame: SummaryFrame,
  resolveOperator: (agent: string) => string,
): ContributionTotals | null {
  // Deduplicated by digest before anything is counted, and this is not
  // defensive tidying: in a joint replay the merged pool legitimately holds the
  // same Result twice — the author's own copy, and the verbatim copy a
  // counterparty received across the boundary (ADR-0009). Summing the pool as
  // given credits that work twice, which is a wrong number arrived at from an
  // entirely correct record.
  const byDigest = new Map<string, Activity>();
  for (const activity of poolIn) byDigest.set(digestOf(activity), activity);
  const pool = [...byDigest.values()];

  const slice = hubChainSlice(pool, frame.periodRule.hub, frame.periodRule.from, frame.periodRule.to);
  if (slice === null) return null;

  const settlements = slice.filter((activity) => payloadOf(activity, "afp:Settlement") !== null);
  const announces = new Map<string, Activity>();
  for (const activity of pool) {
    const task = payloadOf(activity, "afp:Task");
    if (task && typeof task.id === "string") announces.set(task.id, activity);
  }

  const scope = new Set(frame.inputScope.visibility);
  const readable = (activity: Activity): boolean => scope.has(String(activity["afp:visibility"]) as Visibility);

  const rows: { agent: string; share: number; denominator: number }[] = [];
  const unreadable: Record<string, number> = {};
  const inputs = new Set<string>();

  for (const settlement of settlements) {
    inputs.add(digestOf(settlement));
    const object = payloadOf(settlement, "afp:Settlement")!;
    const taskId = String(object["afp:task"] ?? "");
    const announce = announces.get(taskId);
    const thread = announce ? String(announce.context ?? "") : "";
    const results = thread
      ? pool.filter((a) => a.context === thread && payloadOf(a, "afp:Result") !== null && readable(a))
      : [];

    if (results.length === 0) {
      // The hub sequenced a settlement, so the work happened; this computer
      // cannot see the Result that says who did it. Counted, never estimated
      // and never silently skipped (W0.2). Attributed through the settlement's
      // own entries, which the hub published and everyone can read.
      const settled = object["afp:settles"];
      const actors = Array.isArray(settled)
        ? settled.map((entry) => String((entry as Record<string, JsonValue>)?.actor ?? "")).filter(Boolean)
        : [];
      for (const operator of new Set((actors.length ? actors : ["unattributed"]).map(resolveOperator))) {
        unreadable[operator] = (unreadable[operator] ?? 0) + 1;
      }
      continue;
    }

    for (const result of results) {
      inputs.add(digestOf(result));
      rows.push(...creditOf(payloadOf(result, "afp:Result")!));
    }
  }

  // Integer arithmetic only (W0.1): scale every share to the least common
  // multiple of the denominators, exactly as a proposal scales seats.
  const denominator = rows.length ? lcmOf(rows.map((row) => row.denominator)) : 1;
  const credited: Record<string, number> = {};
  for (const row of rows) {
    const operator = resolveOperator(row.agent);
    credited[operator] = (credited[operator] ?? 0) + (row.share * denominator) / row.denominator;
  }

  return { denominator, credited, unreadable, inputs: [...inputs].sort() };
}

/** Decision 3's preimage, stated once and computed identically on both sides. */
export function inputHashOf(inputs: readonly string[]): string {
  return digestOf([...inputs].sort() as unknown as JsonValue);
}

export interface SummarySpec {
  summaryId: string;
  hub: string;
  computedBy: string;
  frame: SummaryFrame;
  totals: ContributionTotals;
  /** Human-readable narration only — never what the arithmetic selects on. */
  period?: { start: string; end: string };
  /** Decision 5: the ratified summary this one corrects, by activity digest. */
  supersedes?: string;
}

/**
 * `Create{afp:ContributionSummary}` — a **draft** until an ADR-0018 round
 * ratifies it (Decision 5). Publishing one is not a claim of authority, and 04
 * never gave anyone that authority in the first place: the object exists so
 * that nobody's arithmetic has to be trusted.
 */
export function contributionSummary(envelope: Envelope, spec: SummarySpec): { [key: string]: JsonValue } {
  const frame: { [key: string]: JsonValue } = {
    "afp:periodRule": {
      "afp:form": spec.frame.periodRule.form,
      "afp:hub": spec.frame.periodRule.hub,
      "afp:from": spec.frame.periodRule.from,
      "afp:to": spec.frame.periodRule.to,
    },
    "afp:inputScope": { "afp:visibility": [...spec.frame.inputScope.visibility] },
    "afp:splitRule": { "afp:form": spec.frame.splitRule.form },
    "afp:vocabulary": spec.frame.vocabulary,
  };

  const object: { [key: string]: JsonValue } = {
    id: spec.summaryId,
    type: "afp:ContributionSummary",
    "afp:hub": spec.hub,
    "afp:computedBy": spec.computedBy,
    "afp:frame": frame,
    "afp:denominator": spec.totals.denominator,
    "afp:entries": Object.keys(spec.totals.credited)
      .sort()
      .map((operator) => ({ "afp:operator": operator, "afp:credited": spec.totals.credited[operator] })),
    "afp:inputHash": inputHashOf(spec.totals.inputs),
  };
  if (spec.period) object["afp:period"] = { start: spec.period.start, end: spec.period.end };
  // Emitted only when there is something to declare (W0.5) — but an empty
  // census and an absent one are different claims, so a computer that read
  // everything says so with an empty list rather than by omission.
  object["afp:unreadable"] = Object.keys(spec.totals.unreadable)
    .sort()
    .map((operator) => ({ "afp:operator": operator, "afp:count": spec.totals.unreadable[operator] }));
  if (spec.supersedes) object["afp:supersedes"] = spec.supersedes;

  return {
    "@context": AFP_CONTEXTS,
    id: envelope.activityId,
    type: "Create",
    actor: envelope.actor,
    to: [...envelope.to],
    published: envelope.published,
    context: envelope.thread,
    "afp:visibility": envelope.visibility,
    ...(envelope.prevActivity !== null ? { "afp:prevActivity": envelope.prevActivity } : {}),
    object,
  };
}
